### DISCUSSION TITLE
Certificate identity for GitHub Actions reusable workflow keyless signing

### DISCUSSION THREAD

**Discussion by sagikazarmark** [sigstore/cosign#2936] [General]
Let's say I have a reusable `workflow A` that builds and signs container images on GitHub Actions using keyless signing.

When I call this from `workflow B` the certificate identity is going to be the workflow ref for `workflow A`.

It doesn't sound like a huge deal, but imagine that reusable workflow is from a third-party: the certificate identity becomes a reference to a third-party workflow instead of my release workflow (`workflow B`).

From a consumer perspective, it's weird the very least.

I'm somewhat convinced this isn't cosign's "fault" as it probably just uses the identity coming from the ID token.

I'd be interested to know if others find this a problem as well or not. If not, why not?

Thanks!

**Comment by Hayden-IO** [comment]
You're correct that the subject of the certificate will be the same, as it's derived from the reusable workflow reference, but there are other claims in the certificate’s extensions that allow you to differentiate between workflows. See https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md for more information.

The OID for workflow is Build Config URI, https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#13614157264118--build-config-uri, whereas reusable workflow is Build Signer URI (and the Subject Alt Name), https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#1361415726419--build-signer-uri. Note that these are newer OIDs that are standardized across CI platforms, and are actively being implemented for GitLab and Buildkite too. There are older OIDs we've marked as deprecated that will be present in previously issued certificates that represent workflow just for GitHub Actions, https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#1361415726414--github-workflow-name-deprecated.

**Reply by sagikazarmark** [reply]
@haydentherapper thanks for your answer!

I probably should have more clear why I'm asking: currently, when you verify the signature, you have to pass a `--certificate-identity` to `cosign verify` which (as far as I understand) needs to be the subject.

Am I wrong in that assumption? Is there a way to use any of those other OIDs instead of the subject?

(Forgive my ignorance, I'm just getting started with cosign)

**Reply by znewman01** [reply]
CC @laurentsimon @asraa for questions about reusable workflows

Lots of options:

```
$ cosign -- verify --help | grep -- '--certificate' | awk '{ print $1 }'
--certificate
--certificate-chain
--certificate-github-workflow-name
--certificate-github-workflow-ref
--certificate-github-workflow-repository
--certificate-github-workflow-sha
--certificate-github-workflow-trigger
--certificate-identity
--certificate-identity-regexp
--certificate-oidc-issuer
--certificate-oidc-issuer-regexp
```

Right now it's not so ergonomic; we're working on that: https://github.com/sigstore/cosign/issues/2691

**Reply by laurentsimon** [reply]
> From a consumer perspective, it's weird the very least.

The reusable workflow is isolated from workflow B. You *are* trusting the workflow A, so security-wise it does make sense that the identity of the signer / builder is workflow A.

This is the "source repo" OID you can use from the new claims https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#13614157264112--source-repository-uri, or the (deprecated, but sill present) https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#1361415726415--github-workflow-repository-deprecated

@sagikazarmark do you have a link to your setup (workflow A)?

**Reply by sagikazarmark** [reply]
@laurentsimon here is my reusable workflow: https://github.com/sagikazarmark/todobackend-go-kit/blob/main/.github/workflows/artifacts.yaml

And this is where I call it from: https://github.com/sagikazarmark/todobackend-go-kit/blob/main/.github/workflows/release.yaml

Currently, it's in the same repo, but I guess it would work the same way with a third-party repo.

> The reusable workflow is isolated from workflow B. You are trusting the workflow A, so security-wise it does make sense that the identity of the signer / builder is workflow A.

I guess it depends on who is "you". From a client perspective, I'm trusting the project that publishes the artifact. So when I try to verify an image built by Dex for example (I'm in the process of implementing image signing there), currently this is what I would see (imaginary example):

```shell
cosign verify --certificate-identity 'https://github.com/some-shady-company/docker-build/.github/workflows/artifacts.yaml@refs/tags/v0' --certificate-oidc-issuer 'https://token.actions.githubusercontent.com' ghcr.io/dexidp/dex:latest
```

Again, from the perspective of someone who wants to use Dex, this at least makes me raise my eyebrows.

> The reusable workflow is isolated from workflow B.

I'm not entirely convinced that's the case. The reusable workflow may run in a separate environment, but `github.workflow_ref` is set to the parent workflow during the reusable workflow run.

Proof: https://github.com/sagikazarmark/todobackend-go-kit/blob/main/.github/workflows/artifacts.yaml#L142

<img width="1285" alt="Screenshot 2023-04-27 at 20 33 07" src="https://user-images.githubusercontent.com/1226384/234959096-5b332324-cf51-4ce7-a32a-338e7e491a64.png">

https://github.com/sagikazarmark/todobackend-go-kit/actions/runs/4658716071

**Reply by laurentsimon** [reply]
Fully understand what you mean. There are two levels of workflows, and you'd rather have the identify of the project's triggered workflow A. If you want this functionality out of the box, you may instead use the SLSA generators https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/generic#readme, in combination with https://github.com/slsa-framework/slsa-verifier to generate provenance. You'll be able to verify as `slsa-verifier verify-image <image> --source-uri github.com/repo/of/workflow/B --source-tag <version>` (the provenance's triggered workflow will be set to `github.com/release.yml`). Similar for the SBOM.

Note: we're working with Syft on an SBOM builder that will scan and generate SLSA provenance for the SBOM, so you'd be able to replace https://github.com/sagikazarmark/todobackend-go-kit/blob/main/.github/workflows/artifacts.yaml#L130 with a call to Syft's SBOM builder (which will be a reusbale workflow itself); and verify it with slsa-verifier.

**Reply by laurentsimon** [reply]
> I'm not entirely convinced that's the case. The reusable workflow may run in a separate environment, but github.workflow_ref is set to the parent workflow during the reusable workflow run.

yes you're right the GH event is passed to the re-usable workflow, but they run on different VMs.

**Reply by sagikazarmark** [reply]
@laurentsimon Thanks for the suggestion! I'm guessing that would still use cosign under the hood somehow?

My goal is to figure out a set of best practices I can apply to a number of open source projects that I'm working on (Dex, Bank-Vaults, etc). Coming from this angle, an ideal solution:

- can be easily integrated into existing build workflows
- easy to understand for contributors
- easy to use for consumers of the project (no "raised eyebrows")

Cosign ticks most of the boxes, except the last one.

Also, I've yet to see a good blog post how all the new supply chain security related tools and standards (sigstore, SLSA, SBOMs, in-toto, etc) work together, because it's not easy to grasp even as an open source maintainer: I can only guess how consumers would feel.

Anyway, this is where I'm coming from. I'm gonna check out those SLSA generators.

Any further thoughts, insights, tips would be greatly appreciated. I'm trying to bring the above mentioned and more projects implement the (current) best practices and once I understand how things are working, I'd be happy to spread the word in the community, write some blog posts, etc.

Thank you!

**Reply by laurentsimon** [reply]
> @laurentsimon Thanks for the suggestion! I'm guessing that would still use cosign under the hood somehow?

It's using Sigstore, yes; and cosign format for container's attestations. Feel free to ask question on the generator repo if you have questions.

**Reply by laurentsimon** [reply]
> Any further thoughts, insights, tips would be greatly appreciated. I'm trying to bring the above mentioned and more projects implement the (current) best practices and once I understand how things are working, I'd be happy to spread the word in the community, write some blog posts, etc.

what's a good contact to chat? Happy to have a chat. I agree not everything is crystal clear today, but I can give you a sense where things are going

**Reply by sagikazarmark** [reply]
Any chance you'll be at OSS NA? I'm on CNCF Slack as "Márk Sági-Kazár".

**Reply by laurentsimon** [reply]
I will! I have issues accessing CNCF slack. I'm on the OpenSSF and Sigstore slack though. Or shoot me an email at `github-username` AT google dot com

re: provenance gen. For containers you would need to use https://github.com/slsa-framework/slsa-github-generator/blob/main/internal/builders/container/README.md. In general, https://github.com/slsa-framework/slsa-github-generator#generation-of-provenance has a list of builders / generators to use.

**Reply by Hayden-IO** [reply]
I'll be around too, happy to chat with y'all :)
